Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemGroupsController.cs

```diff
@@ -0,0 +1,149 @@
+    using GeneralResource = Resources.Areas.Administration.AdministrationGeneral;
+    using Resource = Resources.Areas.Administration.ProblemGroups.ProblemGroupsController;
+    using ViewModelType = OJS.Web.Areas.Administration.ViewModels.ProblemGroup.DetailedProblemGroupViewModel;
+
+    public class ProblemGroupsController : LecturerBaseGridController
+    {
+        private readonly IProblemGroupsDataService problemGroupsData;
+        private readonly IProblemGroupsBusinessService problemGroupsBusiness;
+        private readonly IContestsDataService contestsData;
+
+        public ProblemGroupsController(
+            IOjsData data,
+            IProblemGroupsDataService problemGroupsData,
+            IProblemGroupsBusinessService problemGroupsBusiness,
+            IContestsDataService contestsData)
+            : base(data)
+        {
+            this.problemGroupsData = problemGroupsData;
+            this.problemGroupsBusiness = problemGroupsBusiness;
+            this.contestsData = contestsData;
+        }
+
+        public override IEnumerable GetData() =>
+            this.problemGroupsData
+                .GetAll()
+                .Select(ViewModelType.FromProblemGroup);
+
+        public override object GetById(object id) => this.GetByIdAsNoTracking((int)id);
+
+        public override string GetEntityKeyName() => this.GetEntityKeyNameByType(typeof(ProblemGroup));
+
+        public ActionResult Index() => this.View();
+
+        [HttpPost]
+        public ActionResult Create(
+            [DataSourceRequest] DataSourceRequest request,
+            ViewModelType model)
+        {
+            if (!this.IsModelAndContestValid(model))
+            {
+                return this.GridOperation(request, model);
+            }
+
+            this.BaseCreate(model.GetEntityModel());
+            return this.GridOperation(request, model);
+        }
+
+        [HttpPost]
+        public ActionResult Update([DataSourceRequest] DataSourceRequest request, ViewModelType model)
+        {
+            if (!this.IsModelAndContestValid(model))
+            {
+                return this.GridOperation(request, model);
+            }
+
+            var problemGroup = model.GetEntityModel(this.GetByIdAsNoTracking(model.Id));
+
+            this.BaseUpdate(problemGroup);
+            return this.GridOperation(request, model);
+        }
+
+        [HttpPost]
+        public ActionResult Destroy([DataSourceRequest] DataSourceRequest request, ViewModelType model)
+        {
+            if (!this.IsModelAndContestValid(model))
+            {
+                return this.GridOperation(request, model);
+            }
+
+            var result = this.problemGroupsBusiness.DeleteById(model.Id);
+
+            if (result.IsError)
+            {
+                this.ModelState.AddModelError(string.Empty, result.Error);
+            }
+
+            return this.GridOperation(request, model);
+        }
+
+        public ActionResult Contest(int? id)
```

Comment:
Better naming. Consider using verb-orientated naming instead.

Reply:
I think that the URL would be more user friendly if the name is just Contest: /ProblemGroups/Contest/325

Reply:
You can consider using [attribute routing in this case](https://blogs.msdn.microsoft.com/webdev/2013/10/17/attribute-routing-in-asp-net-mvc-5/)